### PR TITLE
Reduce the amount of startup banner text

### LINKED ISSUE #15644: `-nologo` should be the default
## Steps to reproduce

Launch PowerShell in your favorite terminal.

## Expected behavior

The prompt I've set up in my PowerShell profile.

## Actual behavior

I get the following boilerplate:

```none
PowerShell 7.1.3
Copyright (c) Microsoft Corporation.

https://aka.ms/powershell
Type 'help' to get help.
```

"But Immo, you can turn this off by using `-nologo`". Yes. But I have to do this in every single one of the places that launches the shell: shortcuts in the start menu, pinned task bar entries, the Windows Terminal commands, the VS Code integrated terminal, the Visual Studio terminal, Git GUI clients etc. It's tiring. If there would be a user wide setting, that might be an OK workaround, but every time I set up a new machine I have to add these extra arguments -- just to get a shell that is ready for use.

The year is 2021. We're embracing command line stuff and simple text files for the modern dev environment (really, we're just rediscovering Unix philosophy, but that's a different rant 😄).  It seems the default shell behavior should optimize for this kind of user, rather than forcing this boilerplate on everyone that arguably adds zero value to their experience.

### PR DESCRIPTION
**Pull request by rkeithhill**
# PR Summary

This PR does not set -nologo as the default, instead it removes the
copyright line, the two "how to get help" lines and two blank lines
around the "how to get help" lines.

The following is the resulting startup banner when there is no update available OR the user has disabled updating checking via the `POWERSHELL_UPDATECHECK` environment variable:

image

compare this with the current startup banner:

image

When there is an update available AND the user has not disabled update checking, the reduced startup banner looks like this:

image

If we decide that we want to keep the `Type 'help' to get help.` text, I propose something like this:

image

image

BTW here are some other startup banners for comparison:

image

As you can see, bash and nodejs have no startup banner text at all and Python's is a bit wordy.  :-)

## PR Context

Addresses #15644

Regarding documentation impact, I think this would only impact images/text blocks where the PowerShell startup banner is shown.  But what I'm seeing so far in the online help is images of Windows PowerShell startup banner text.  @sdwheeler are there areas of the online help where PowerShell v6/7 startup banner text is shown?

This PR currently leaves the profile timing text as-is.

The PR is meant to give folks a way to test drive a reduced startup banner.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/8476
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Comment by iSazonov** [comment]
Convert to draft until PowerShell Committee conclusion.

**Comment by sdwheeler** [comment]
There are 3 articles:
- [about_pwsh](https://docs.microsoft.com/powershell/module/microsoft.powershell.core/about/about_pwsh) - documents `-NoLogo` but does not show an example
- [about_environment_variables](https://docs.microsoft.com/powershell/module/microsoft.powershell.core/about/about_environment_variables) - documents `$env:POWERSHELL_UPDATECHECK` but not really affected by this PR
- [about_update_notifications](https://docs.microsoft.com/powershell/module/microsoft.powershell.core/about/about_update_notifications) - documents how Update Notification works but not really affected by this PR

**Comment by SteveL-MSFT** [comment]
@rkeithhill would you mind adding a test?

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this, we are ok taking this PR with just the `PowerShell <version>` logo and getting user feedback.  We still need to follow-up with Microsoft to ensure we are ok without the copyright notice, but that's something we can address later and doesn't block this PR.

**Review by daxian-dbw — APPROVED** [review]

**Comment by rkeithhill** [comment]
@SteveL-MSFT I can add look at adding a test.  Are there any existing tests for the startup banner text?

File: src/Microsoft.PowerShell.ConsoleHost/resources/ManagedEntranceStrings.resx [hunk 1]
```diff
@@ -118,11 +118,7 @@
     <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
   </resheader>
   <data name="ShellBannerNonWindowsPowerShell" xml:space="preserve">
-    <value>PowerShell {0}
-Copyright (c) Microsoft Corporation.
-
-https://aka.ms/powershell
-Type 'help' to get help.</value>
+    <value>PowerShell {0}</value>
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/resources/ManagedEntranceStrings.resx:121]:
```suggestion
    <value>PowerShell {0}
    </value>
```

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/UpdatesNotification.cs [hunk 2]
```diff
@@ -126,6 +126,7 @@ internal static void ShowUpdateNotification(PSHostUserInterface hostUI)
 
                 string notificationMsg = string.Format(CultureInfo.CurrentCulture, notificationMsgTemplate, releaseTag, notificationColor, resetColor, line2Padding, line3Padding);
 
+                hostUI.WriteLine();
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/UpdatesNotification.cs:129]:
```suggestion
```

**Comment by iSazonov** [comment]
@rkeithhill We have xUnit tests in test_CommandLineParser.cs. You could add new tests in the file.

**Comment by iSazonov** [comment]
> @PowerShell/powershell-committee reviewed this, we are ok taking this PR with just the `PowerShell <version>` logo and getting user feedback. We still need to follow-up with Microsoft to ensure we are ok without the copyright notice, but that's something we can address later and doesn't block this PR.

There is a related problems with "<version>". I'd ask the team and PowerShell Committee to look https://github.com/PowerShell/PowerShell/issues/15607#issuecomment-985368909 so that we could continue with #15603.

File: src/Microsoft.PowerShell.ConsoleHost/resources/ManagedEntranceStrings.resx (hunk 1 above)
Comment (rkeithhill) [reply src/Microsoft.PowerShell.ConsoleHost/resources/ManagedEntranceStrings.resx:121]: I don't think we want a blank line after `PowerShell 7.3.0`.   But the changes you are suggesting are to provide just that.

Comment (iSazonov) [reply src/Microsoft.PowerShell.ConsoleHost/resources/ManagedEntranceStrings.resx:121]: If so we need to remove `hostUI.WriteLine();`

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/UpdatesNotification.cs (hunk 2 above)
Comment (daxian-dbw) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/UpdatesNotification.cs:129]: We want an empty line when we need to write out update notification message. Otherwise, no blank line. So, the existing change is desired.

**Comment by rkeithhill** [comment]
@SteveL-MSFT  I think the best place for tests is in `ConsoleHost.Tests.ps1`.  However, I'm a bit stumped on how to test this.  All of the other tests in this file appear to use either `-File` or `-Command` in order to run pwsh, execute script, exit and capture the output.  In those cases, no startup banner is output.  If you run without those parameters, you get the startup banner in the new pwsh session but that session doesn't exit and getting the startup banner text from that session isn't exactly straightforward.

**Comment by iSazonov** [comment]
@rkeithhill I still think we can place tests in test_CommandLineParser.cs
If you want a script test you could look how we use New-TestHost in tests.

**Comment by rkeithhill** [comment]
@iSazonov I don't see how `New-TestHost` helps.  That host will have its own host startup text (likely none at all) and would not be testing `Microsoft.PowerShell.ConsoleHost` start up, which is what displays the modified startup banner text.  This is a similar situation to the VSCode extension host which has its own startup banner text.  And AFAICT, `test_CommandLineParser.cs` is testing just the parsing of command line parameters and not any output from actually starting PowerShell.  This PR doesn't change any of the PowerShell.exe parameter parsing behavior.

**Comment by iSazonov** [comment]
@rkeithhill I say about `emulation`. You could call `ShowErrorHelpBanner`. It has `hostUI` parameter (you can use our test console) and `bannerText` (you could use `ShellBannerNonWindowsPowerShell` format string for it).
Otherwise you have to grab the real pwsh output but I don't know how to do it without a test hook.

**Comment by iSazonov** [comment]
In other words, it is easier to test the banner itself than the host.

**Comment by adityapatwardhan** [comment]
Waiting on test.

**Comment by rkeithhill** [comment]
@adityapatwardhan I need some guidance on how to test this?  Using a different host i.e., test host isn't really testing **the** msh host's startup banner text.  And as I mentioned above, testing pwsh in a way that executes a command/file does not output banner text at all.  And while running pwsh.exe without -command/-file outputs startup banner text, it does so in a different process.

**Comment by adityapatwardhan** [comment]
I have a "creative" solution. :P. Dont know if it will be stable enough in CI.

```powershell
$p = Start-Process -FilePath pwsh.exe -RedirectStandardOutput .\op.txt -PassThru -NoNewWindow ; start-sleep -Seconds 3 ; Stop-Process $p; Get-Content .\op.txt

PowerShell 7.2.1
Copyright (c) Microsoft Corporation.

https://aka.ms/powershell
Type 'help' to get help.
```

File: test/powershell/Host/ConsoleHost.Tests.ps1 [hunk 3]
```diff
@@ -833,6 +833,58 @@ namespace StackTest {
...
+            $outputPath = Join-Path $env:temp "StartupBanner-${Pid}.txt"
+
+            # Not testing update notification banner text here
+            $origUpdateCheckVal = $env:POWERSHELL_UPDATECHECK
+            $env:POWERSHELL_UPDATECHECK = "Off"
+
+            $spArgs = @{
+                FilePath = $powershell
+                ArgumentList = @("-NoProfile")
+                RedirectStandardOutput = $outputPath
```
Comment (iSazonov) [inline test/powershell/Host/ConsoleHost.Tests.ps1:849]: Can we use `RedirectStandardInput` and send `Exit` command so that avoid Start-Sleep?

Comment (rkeithhill) [reply test/powershell/Host/ConsoleHost.Tests.ps1:849]: I think so.  I'll still need to use Start-Sleep in a `while` loop to wait for `!$p.HasExited` but this would be more robust.  Thanks.

Comment (rkeithhill) [reply test/powershell/Host/ConsoleHost.Tests.ps1:849]: Check that, I can use `$p.WaitForExit(5000)`?

Comment (iSazonov) [reply test/powershell/Host/ConsoleHost.Tests.ps1:849]: Yes, you can. Or we could use our helper:
```
Wait-UntilTrue -sb { $process.HasExited } -TimeoutInMilliseconds 5000 -IntervalInMilliseconds 1000 | Should -BeTrue
```

File: src/Microsoft.PowerShell.ConsoleHost/resources/ManagedEntranceStrings.resx [hunk 4]
```diff
@@ -383,7 +379,7 @@ All parameters are case-insensitive.</value>
 
 -NoLogo | -nol
 
-    Hides the copyright banner at startup of interactive sessions.
+    Hides the banner text at startup of interactive sessions.
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/resources/ManagedEntranceStrings.resx:382]: @rkeithhill  Please open doc issue.
/cc @sdwheeler.

File: test/powershell/Host/ConsoleHost.Tests.ps1 [hunk 5]
```diff
@@ -833,6 +833,60 @@ namespace StackTest {
...
+            # Restore original value of $env:POWERSHELL_UPDATECHECK
+            if ($origUpdateCheckVal) {
+                $env:POWERSHELL_UPDATECHECK = $origUpdateCheckVal
+            }
+            else {
+                Remove-Item Env:\POWERSHELL_UPDATECHECK -Force
+            }
+        }
+        BeforeEach {
+            Remove-Item $outputPath -Force -ErrorAction Ignore
```
Comment (iSazonov) [inline test/powershell/Host/ConsoleHost.Tests.ps1:868]: We need to remove `$inputPath`  too. And in AfterAll too.

Comment (rkeithhill) [reply test/powershell/Host/ConsoleHost.Tests.ps1:871]: I've add script to remove the stdin file in AfterAll but I do not want to remove it in BeforeEach.  It is created once in BeforeAll and each test uses the same stdin text (`exit`).  There's no point in re-creating that same text in the stdin file for each test.

**Review by iSazonov — APPROVED** [review]

File: src/Microsoft.PowerShell.ConsoleHost/resources/ManagedEntranceStrings.resx (hunk 4 above)
Comment (rkeithhill) [reply src/Microsoft.PowerShell.ConsoleHost/resources/ManagedEntranceStrings.resx:382]: Done. https://github.com/MicrosoftDocs/PowerShell-Docs/issues/8476

**Comment by rkeithhill** [comment]
FYI, I'm trying to sort out the Linux Pester test failure but the line numbers the log shows seems odd.  Anyway, trying to build & run my tests on WSL/Ubuntu to see if I can repro locally.

**Comment by rkeithhill** [comment]
@iSazonov @adityapatwardhan Any idea why I would be getting this output from the `Azure Pipelines / PowerShell-CI-macos` & `Azure Pipelines / PowerShell-CI-linux` tests?

```
WARNING: pwsh output is: 'PowerShell 7.3.0-preview.2\nPS /home/vsts/work/1/s> 1hexit\n1l'
```
What's with the `1h`, extra newline and final `1l`?  I suppose I need to disable ANSI for the new pwsh process?  Now I'm wanting a `-NoAnsi` switch to pwsh.exe.

**Comment by iSazonov** [comment]
@rkeithhill I could set env variable TERM=dumb to disable VT mode.

**Comment by rkeithhill** [comment]
@iSazonov Thanks but I've found something that works.  I'm not keen to set the TERM=dumb on the Pester process side.  And doing it in the new pwsh process isn't much different than my current fix which is to first execute: `$PSStyle.OutputRendering = 'PlainText'`.

**Comment by iSazonov** [comment]
TERM=dumb turns off DECCKM too and simplify the test code.

File: test/powershell/Host/ConsoleHost.Tests.ps1 [hunk 6]
```diff
@@ -833,6 +833,64 @@ namespace StackTest {
...
+        }
+        It "Displays expected startup banner text by default" {
+            $process = Start-Process @spArgs
+            Wait-UntilTrue -sb { $process.HasExited } -TimeoutInMilliseconds 5000 -IntervalInMilliseconds 250 | Should -BeTrue
+
+            $out = @(Get-Content $outputPath)
+            $out.Count | Should -BeExactly 2
```
Comment (iSazonov) [inline test/powershell/Host/ConsoleHost.Tests.ps1:880]: BeExactly is used for case-sensetive string comparisons. 
```suggestion
            $out.Count | Should -Be 2
```

**Review by iSazonov — APPROVED** [review]

**Comment by rkeithhill** [comment]
> TERM=dumb turns off DECCKM too and simplify the test code.

@iSazonov  Done.  Thanks.

**Comment by rkeithhill** [comment]
@adityapatwardhan This PR is ready to go (merge).

File: test/powershell/Host/ConsoleHost.Tests.ps1 [hunk 7]
```diff
@@ -833,6 +833,64 @@ namespace StackTest {
...
+            $inputPath  = "Temp:\StartupBannerTest-Input.txt"
+            "exit" > $inputPath
+
+            # Not testing update notification banner text here
+            $oldPowerShellUpdateCheck = $env:POWERSHELL_UPDATECHECK
+            $env:POWERSHELL_UPDATECHECK = "Off"
+
+            # Set TERM to "dumb" to avoid DECCKM codes in the output
+            $oldTERM = $env:TERM
+            $env:TERM = "dumb"
```
Comment (adityapatwardhan) [inline test/powershell/Host/ConsoleHost.Tests.ps1:849]: Should this be done only on non-Windows?

Comment (rkeithhill) [reply test/powershell/Host/ConsoleHost.Tests.ps1:849]: It probably only **needs** to be done non-WIndows since the DECCKM failures I saw where on mac/linux.  That said, I don't believe it hurts to set TERM="dumb" on Windows.

**Review by adityapatwardhan — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
@rkeithhill Thank you for your contribution!

**Comment by DHowett** [comment]
I just noticed this. _Thank you_.
